llvm/lib/Transforms/Utils/SimplifyCFG.cpp

```diff
@@ -6552,6 +6552,20 @@ static bool SwitchToLookupTable(SwitchInst *SI, IRBuilder<> &Builder,
     // Note: We call removeProdecessor later since we need to be able to get the
     // PHI value for the default case in case we're using a bit mask.
   } else {
+    const APInt *MaskC;
+    if (UseSwitchConditionAsTableIndex &&
+        match(TableIndex, m_And(m_Value(), m_APInt(MaskC)))) {
```

Comment:
This seems unfortunately specific. For example if we replace `x & 15` with `x % 42` or so, it will no longer work, even though the basic idea is the same. We could use something like computeConstantRange() to generalize this.